File: 1/inversions.js
/**
 * Chris Amatangelo
 * Programming Question - 1
 */
var fs = require('fs');

var lines = fs.readFileSync('./IntegerArray.txt').toString().split("\n").filter(Boolean).map(Number);

function mergeAndCountSplitInv(arr, left, right) {
  var i = 0, j = 0, count = 0;

  for (var k = 0; k < arr.length; k++) {
    if (left[i] < right[j]) {
      arr[k] = left[i];
      i++;
    } else if (right[j] < left[i]) {
      arr[k] = right[j];
      count += left.length - i;
      j++;
    } else if (j == right.length) {
      arr[k] = left[i];
      i++;
    } else if (i == left.length) {
      arr[k] = right[j];
      j++;
    }
  }
  return count;
}

var x = 0;
var y = 0;
var z = 0;

function sortAndCount(arr) {
  if (arr.length < 2) return 0;

  var n = Math.floor(arr.length / 2);
  var right = arr.slice(n);
  var left  = arr.slice(0, n); 

  x = sortAndCount(left);
  y = sortAndCount(right);
  z += mergeAndCountSplitInv(arr, left, right);

  return z;
}

console.log( sortAndCount(lines) );

File: 1/inversions.py
import math

# with open('IntegerArray.txt', 'r') as f:
# lines = [line.rstrip('\n') for line in open('IntegerArray.txt')]
lines = [6,5,4,3,2,1] # => 15

# len(arr)
def count(arr, n):
  if n == 1:
    return 0
  else:
    print 'Right'
    print int(n/2)
    count(int(n/2))
    print 'Left'
    print (n + 2 // 2) // 2
    count((n + 2 // 2) // 2)
    # x = count(arr[0:int(math.floor(n / 2))], int(n/2))
    # y = count(arr[int(math.floor(n / 2)):len(arr)], (n + 2 // 2) // 2)
    # y = count(arr[], (n + 2 // 2) // 2, int(n/2))
    # y = count
  # return x + y

print count(lines, len(lines))


# roughly 5 billion possible inversions

# for line in lines:
#   print line
